Is it possible to migrate existing VM to another target node?

System details:

- Proxmox VE 8.2.2
- Terraform v1.8.4
- terraform-provider-proxmox v3.0.1-rc2

The existing VM was successfully created by the Terraform provider, on node `pve10`.

There is a task to migrate a VM to another node.

I assume that `target_node` parameter is responsible for this.

So changing this node to another (`pve10 => pve11`) and doing `terraform apply` fails with the following result:

```log
proxmox_vm_qemu.cloud_vm_from_packer_template: Modifying... [id=pve10/qemu/11]
2024-06-04T09:41:23.803+0300 [INFO]  Starting apply for proxmox_vm_qemu.cloud_vm_from_packer_template
2024-06-04T09:41:23.809+0300 [DEBUG] proxmox_vm_qemu.cloud_vm_from_packer_template: applying the planned Update change
2024-06-04T09:41:29.856+0300 [INFO]  provider.terraform-provider-proxmox: 2024/06/04 09:41:29 500 Configuration file 'nodes/pve11/qemu-server/11.conf' does not exist: timestamp="2024-06-04T09:41:29.855+0300"
2024-06-04T09:41:29.859+0300 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-06-04T09:41:29.859+0300 [ERROR] plugin.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-06-04T09:41:29.859+0300 [DEBUG] provider: plugin process exited: path=.terraform/providers/localhost/telmate/proxmox/3.0.2/linux_amd64/terraform-provider-proxmox pid=11292 error="exit status 1"
2024-06-04T09:41:29.878+0300 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2024-06-04T09:41:29.878+0300 [ERROR] vertex "proxmox_vm_qemu.cloud_vm_from_packer_template" error: Plugin did not respond
╷
│ Error: Plugin did not respond
│ 
│   with proxmox_vm_qemu.cloud_vm_from_packer_template,
│   on main.tf line 39, in resource "proxmox_vm_qemu" "cloud_vm_from_packer_template":
│   39: resource "proxmox_vm_qemu" "cloud_vm_from_packer_template" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
2024-06-04T09:41:29.887+0300 [DEBUG] provider: plugin exited

```

In debug log there are such errors:

```log
...
2024/06/04 09:56:52 >>>>>>>>>> REQUEST:
GET /api2/json/cluster/ha/resources/11 HTTP/1.1
... 
2024/06/04 09:56:52 <<<<<<<<<< RESULT:
HTTP/1.1 500 no such resource 'vm:11'
...
2024/06/04 09:57:03 >>>>>>>>>> REQUEST:
GET /api2/json/nodes/pve11/qemu/11/config HTTP/1.1
...
{"data":null}
2024/06/04 09:57:06 500 Configuration file 'nodes/pve11/qemu-server/11.conf' does not exist
...
```

--- a/docs/resources/lxc.md
+++ b/docs/resources/lxc.md
@@ -267,6 +267,7 @@ The following arguments may be optionally defined when using this resource:
 * `unique` - A boolean that determines if a unique random ethernet address is assigned to the container.
 * `unprivileged` - A boolean that makes the container run as an unprivileged user. Default is `false`.
 * `vmid` - A number that sets the VMID of the container. If set to `0`, the next available VMID is used. Default is `0`.
+* `current_node` __(computed)__ - A string that shows on which node the LXC guest exists.|
 
 ## Attribute Reference
 

--- a/docs/resources/vm_qemu.md
+++ b/docs/resources/vm_qemu.md
@@ -92,7 +92,8 @@ The following arguments are supported in the top level resource block.
 | Argument                      | Type     | Default Value        | Description |
 | ----------------------------- | -------- | -------------------- | ----------- |
 | `name`                        | `str`    |                      | **Required** The name of the VM within Proxmox. |
-| `target_node`                 | `str`    |                      | **Required** The name of the Proxmox Node on which to place the VM. |
+| `target_node`                 | `str`    |                      | The name of the PVE Node on which to place the VM.|
+| `target_nodes`                | `str`    |                      | A list of PVE node names on which to place the VM.|
 | `vmid`                        | `int`    | `0`                  | The ID of the VM in Proxmox. The default value of `0` indicates it should use the next available ID in the sequence. |
 | `desc`                        | `str`    |                      | The description of the VM. Shows as the 'Notes' field in the Proxmox GUI. |
 | `define_connection_info`      | `bool`   | `true`               | Whether to let terraform define the (SSH) connection parameters for preprovisioners, see config block below. |
@@ -145,6 +146,7 @@ The following arguments are supported in the top level resource block.
 | `skip_ipv4`                   | `bool`   | `false`              | Tells proxmox that acquiring an IPv4 address from the qemu guest agent isn't required, it will still return an ipv4 address if it could obtain one. Useful for reducing retries in environments without ipv4.|
 | `skip_ipv6`                   | `bool`   | `false`              | Tells proxmox that acquiring an IPv6 address from the qemu guest agent isn't required, it will still return an ipv6 address if it could obtain one. Useful for reducing retries in environments without ipv6.|
 | `agent_timeout`               | `int`    | `60`                 | Timeout in seconds to keep trying to obtain an IP address from the guest agent one we have a connection. |
+| `current_node`                | `string` |                      | **Computed** The current node of the Qemu guest is on.|
 
 ### VGA Block
 

--- a/go.mod
+++ b/go.mod
@@ -5,7 +5,7 @@ go 1.21
 toolchain go1.21.0
 
 require (
-	github.com/Telmate/proxmox-api-go v0.0.0-20250301154022-c19968ff06eb
+	github.com/Telmate/proxmox-api-go v0.0.0-20250312221820-b57bd2eabcaa
 	github.com/google/uuid v1.6.0
 	github.com/hashicorp/go-cty v1.4.1-0.20200414143053-d3edf31b6320
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.34.0

--- a/proxmox/Internal/resource/guest/node/schema.go
+++ b/proxmox/Internal/resource/guest/node/schema.go
@@ -11,10 +11,12 @@ import (
 const (
 	RootNode  string = "target_node"
 	RootNodes string = "target_nodes"
+	Computed  string = "current_node"
 )
 
 func SchemaNode(s schema.Schema, guestType string) *schema.Schema {
 	s.Type = schema.TypeString
+	s.Optional = true
 	s.Description = "The node the " + guestType + " guest goes to."
 	s.ValidateDiagFunc = func(i interface{}, path cty.Path) diag.Diagnostics {
 		v, ok := i.(string)
@@ -36,11 +38,12 @@ func SchemaNode(s schema.Schema, guestType string) *schema.Schema {
 	return &s
 }
 
-func SchemaNodes() *schema.Schema {
+func SchemaNodes(guestType string) *schema.Schema {
 	return &schema.Schema{
-		Type:        schema.TypeList, // TODO should be `schema.TypeSet`
-		Optional:    true,
-		Description: "A list of nodes the qemu guest goes to.",
+		Type:          schema.TypeSet,
+		Optional:      true,
+		Description:   "A list of nodes the " + guestType + " guest may be placed on.",
+		ConflictsWith: []string{RootNode},
 		Elem: &schema.Schema{
 			Type: schema.TypeString,
 			ValidateDiagFunc: func(i interface{}, path cty.Path) diag.Diagnostics {
@@ -60,4 +63,11 @@ func SchemaNodes() *schema.Schema {
 				}
 				return nil
 			}}}
+}
+
+func SchemaComputed(guestType string) *schema.Schema {
+	return &schema.Schema{
+		Type:        schema.TypeString,
+		Computed:    true,
+		Description: "The node the " + guestType + " guest is currently on."}
 }

--- a/proxmox/Internal/resource/guest/node/sdk.go
+++ b/proxmox/Internal/resource/guest/node/sdk.go
@@ -1,0 +1,49 @@
+package node
+
+import (
+	"math/rand"
+	"time"
+
+	pveAPI "github.com/Telmate/proxmox-api-go/proxmox"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func SdkUpdate(d *schema.ResourceData, current pveAPI.NodeName) pveAPI.NodeName {
+	if node, ok := d.GetOk(RootNode); ok {
+		return pveAPI.NodeName(node.(string))
+	}
+	nodes := d.Get(RootNodes).(*schema.Set).List()
+	currentNode := string(current)
+	if len(nodes) == 1 {
+		if currentNode != nodes[0].(string) {
+			return pveAPI.NodeName(nodes[0].(string))
+		}
+		return current
+	}
+	if inArray(nodes, currentNode) {
+		return current
+	}
+	randomIndex := rand.New(rand.NewSource(time.Now().UnixNano())).Intn(len(nodes))
+	return pveAPI.NodeName(nodes[randomIndex].(string))
+}
+
+func SdkCreate(d *schema.ResourceData) pveAPI.NodeName {
+	if node, ok := d.GetOk(RootNode); ok {
+		return pveAPI.NodeName(node.(string))
+	}
+	nodes := d.Get(RootNodes).(*schema.Set).List()
+	if len(nodes) == 1 {
+		return pveAPI.NodeName(nodes[0].(string))
+	}
+	randomIndex := rand.New(rand.NewSource(time.Now().UnixNano())).Intn(len(nodes))
+	return pveAPI.NodeName(nodes[randomIndex].(string))
+}
+
+func inArray(nodes []interface{}, current string) bool {
+	for i := range nodes {
+		if current == nodes[i].(string) {
+			return true
+		}
+	}
+	return false
+}

--- a/proxmox/Internal/resource/guest/node/terraform.go
+++ b/proxmox/Internal/resource/guest/node/terraform.go
@@ -1,0 +1,21 @@
+package node
+
+import (
+	pveAPI "github.com/Telmate/proxmox-api-go/proxmox"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func Terraform(d *schema.ResourceData, currentNode pveAPI.NodeName) {
+	current := string(currentNode)
+	d.Set(Computed, current)
+	if _, ok := d.GetOk(RootNode); ok {
+		d.Set(RootNode, current)
+		return
+	}
+	nodes := d.Get(RootNodes).(*schema.Set).List()
+	if inArray(nodes, current) {
+		d.Set(RootNodes, nodes)
+		return
+	}
+	d.Set(RootNodes, []interface{}{current})
+}

--- a/proxmox/resource_lxc.go
+++ b/proxmox/resource_lxc.go
@@ -422,9 +422,9 @@ func resourceLxc() *schema.Resource {
 					Type: schema.TypeString,
 				},
 			},
-			node.RootNode: node.SchemaNode(schema.Schema{
-				Required: true, ForceNew: true}, "lxc"),
-			vmID.Root: vmID.Schema(),
+			node.Computed: node.SchemaComputed("qemu"),
+			node.RootNode: node.SchemaNode(schema.Schema{}, "lxc"),
+			vmID.Root:     vmID.Schema(),
 		},
 		Timeouts: resourceTimeouts(),
 	}
@@ -485,8 +485,6 @@ func resourceLxcCreate(ctx context.Context, d *schema.ResourceData, meta interfa
 	config.Unique = d.Get("unique").(bool)
 	config.Unprivileged = d.Get("unprivileged").(bool)
 
-	targetNode := pveSDK.NodeName(d.Get(node.RootNode).(string))
-
 	// proxmox api allows multiple network sets,
 	// having a unique 'id' parameter foreach set
 	networks := d.Get("network").([]interface{})
@@ -513,6 +511,7 @@ func resourceLxcCreate(ctx context.Context, d *schema.ResourceData, meta interfa
 
 	setGuestID := d.Get(vmID.Root).(int)
 
+	targetNode := node.SdkCreate(d)
 	var diags diag.Diagnostics
 	var err error
 	var vmr *pveSDK.VmRef
@@ -820,7 +819,7 @@ func _resourceLxcRead(ctx context.Context, d *schema.ResourceData, meta interfac
 		return err
 	}
 	d.SetId(resourceId(vmr.Node(), "lxc", vmr.VmId()))
-	d.Set(node.RootNode, vmr.Node())
+	node.Terraform(d, vmr.Node())
 
 	// Read Features
 	defaultFeatures := d.Get("features").(*schema.Set)

--- a/proxmox/resource_vm_qemu.go
+++ b/proxmox/resource_vm_qemu.go
@@ -7,7 +7,6 @@ import (
 	"errors"
 	"fmt"
 	"log"
-	"math/rand"
 	"net"
 	"net/url"
 	"path"
@@ -142,9 +141,9 @@ func resourceVmQemu() *schema.Resource {
 				Default:     defaultDescription,
 				Description: "The VM description",
 			},
-			node.RootNode: node.SchemaNode(schema.Schema{
-				Optional: true}, "qemu"),
-			node.RootNodes: node.SchemaNodes(),
+			node.Computed:  node.SchemaComputed("qemu"),
+			node.RootNode:  node.SchemaNode(schema.Schema{ConflictsWith: []string{node.RootNodes}}, "qemu"),
+			node.RootNodes: node.SchemaNodes("qemu"),
 			"bios": {
 				Type:             schema.TypeString,
 				Optional:         true,
@@ -739,27 +738,8 @@ func resourceVmQemuCreate(ctx context.Context, d *schema.ResourceData, meta inte
 
 	forceCreate := d.Get("force_create").(bool)
 
-	targetNodesRaw := d.Get(node.RootNodes).([]interface{})
-	var targetNodes = make([]string, len(targetNodesRaw))
-	for i, raw := range targetNodesRaw {
-		targetNodes[i] = raw.(string)
-	}
-
-	var targetNode pveSDK.NodeName
-
-	if len(targetNodes) == 0 {
-		targetNode = pveSDK.NodeName(d.Get(node.RootNode).(string))
-	} else {
-		targetNode = pveSDK.NodeName(targetNodes[rand.Intn(len(targetNodes))])
-	}
-
-	if targetNode == "" {
-		return diag.FromErr(fmt.Errorf("VM name (%s) has no target node! Please use "+node.RootNode+" or "+node.RootNodes+" to set a specific node! %v", guestName, targetNodes))
-	}
 	if duplicateVmr != nil && forceCreate {
 		return diag.FromErr(fmt.Errorf("duplicate VM name (%s) with vmId: %d. Set force_create=false to recycle", guestName, duplicateVmr.VmId()))
-	} else if duplicateVmr != nil && duplicateVmr.Node() != targetNode {
-		return diag.FromErr(fmt.Errorf("duplicate VM name (%s) with vmId: %d on different %s=%s", guestName, duplicateVmr.VmId(), node.RootNodes, duplicateVmr.Node()))
 	}
 
 	vmr := duplicateVmr
@@ -768,6 +748,7 @@ func resourceVmQemuCreate(ctx context.Context, d *schema.ResourceData, meta inte
 	var err error
 
 	if vmr == nil { // Create new VM
+		targetNode := node.SdkCreate(d)
 		config.Node = &targetNode
 
 		var guestID *pveSDK.GuestID
@@ -861,6 +842,7 @@ func resourceVmQemuCreate(ctx context.Context, d *schema.ResourceData, meta inte
 	} else { // Forcefully update an existing VM
 		log.Printf("[DEBUG][QemuVmCreate] recycling VM vmId: %d", vmr.VmId())
 
+		targetNode := node.SdkUpdate(d, vmr.Node())
 		client.StopVm(ctx, vmr)
 
 		rebootRequired, err = config.Update(ctx, false, vmr, client)
@@ -871,7 +853,7 @@ func resourceVmQemuCreate(ctx context.Context, d *schema.ResourceData, meta inte
 		}
 
 	}
-	d.SetId(resourceId(targetNode, "qemu", vmr.VmId()))
+	d.SetId(resourceId(vmr.Node(), "qemu", vmr.VmId()))
 	logger.Debug().Int(vmID.Root, int(vmr.VmId())).Msgf("Set this vm (resource Id) to '%v'", d.Id())
 
 	// give sometime to proxmox to catchup
@@ -925,15 +907,9 @@ func resourceVmQemuUpdate(ctx context.Context, d *schema.ResourceData, meta inte
 	vga := d.Get("vga").(*schema.Set)
 	qemuVgaList := vga.List()
 
-	d.Partial(true)
-	if d.HasChange(node.RootNode) {
-		// Update target node when it must be migrated manually. Don't if it has been migrated by the proxmox high availability system.
-		vmr.SetNode(d.Get(node.RootNode).(string))
-	}
-	d.Partial(false)
-
 	config := pveSDK.ConfigQemu{
 		Name:        d.Get("name").(string),
+		Node:        util.Pointer(node.SdkUpdate(d, vmr.Node())),
 		CPU:         cpu.SDK(d),
 		Description: util.Pointer(d.Get("desc").(string)),
 		Pool:        util.Pointer(pveSDK.PoolName(d.Get(pool.Root).(string))),
@@ -1164,46 +1140,25 @@ func resourceVmQemuRead(ctx context.Context, d *schema.ResourceData, meta interf
 	// that indicates the VM does not exist. We indicate that to terraform
 	// by calling a SetId("")
 
-	// loop through all virtual servers...?
-	var targetNodeVMR pveSDK.NodeName
-	targetNodesRaw := d.Get(node.RootNodes).([]interface{})
-	targetNodes := make([]pveSDK.NodeName, len(targetNodesRaw))
-	for i := range targetNodesRaw {
-		targetNodes[i] = pveSDK.NodeName(targetNodesRaw[i].(string))
-	}
-
-	if len(targetNodes) == 0 {
-		_, err = client.GetVmInfo(ctx, vmr)
-		if err != nil {
-			logger.Debug().Int(vmID.Root, int(guestID)).Err(err).Msg("failed to get vm info")
+	// not sure if we want to set the id to "" if the vm does not exist
+	// as it will cause terraform to delete the resource
+	// and it could be unavailable due to permission issues
+	// when we are `root@pam` then we can do it as we can see all vms
+	_, err = client.GetVmInfo(ctx, vmr)
+	if err != nil {
+		if err.Error() == "vm '"+vmr.VmId().String()+"' not found" {
+			logger.Debug().Int(vmID.Root, int(guestID)).Err(err).Msg("vm does not exist")
 			d.SetId("")
 			return nil
 		}
-		targetNodeVMR = vmr.Node()
-	} else {
-		for _, targetNode := range targetNodes {
-			vmr.SetNode(targetNode.String())
-			_, err = client.GetVmInfo(ctx, vmr)
-			if err != nil {
-				d.SetId("")
-			}
-
-			d.SetId(resourceId(vmr.Node(), "qemu", vmr.VmId()))
-			logger.Debug().Any("Setting node id to", d.Get(vmr.Node().String()))
-			targetNodeVMR = targetNode
-		}
-	}
-
-	if targetNodeVMR == "" {
-		logger.Debug().Int(vmID.Root, int(guestID)).Err(err).Msg("failed to get vm info")
-		d.SetId("")
-		return nil
+		return append(diags, diag.FromErr(err)...)
 	}
 
 	config, err := pveSDK.NewConfigQemuFromApi(ctx, vmr, client)
 	if err != nil {
 		return diag.FromErr(err)
 	}
+	node.Terraform(d, vmr.Node())
 
 	var ciDisk bool
 	if config.Disks != nil {

--- a/proxmox/util.go
+++ b/proxmox/util.go
@@ -357,12 +357,14 @@ func resourceDataToFlatValues(d *schema.ResourceData, resource *schema.Resource)
 		case schema.TypeFloat:
 			flatValues[key] = d.Get(key).(float64)
 		case schema.TypeSet:
-			values, _ := schemaSetToFlatValues(d.Get(key).(*schema.Set), value.Elem.(*schema.Resource))
-			flatValues[key] = values
+			if _, ok := value.Elem.(*schema.Schema); ok {
+				flatValues[key] = value.Elem.(*schema.Schema)
+			} else {
+				values, _ := schemaSetToFlatValues(d.Get(key).(*schema.Set), value.Elem.(*schema.Resource))
+				flatValues[key] = values
+			}
 		case schema.TypeList:
-			_, ok := value.Elem.(*schema.Schema)
-
-			if ok {
+			if _, ok := value.Elem.(*schema.Schema); ok {
 				flatValues[key] = value.Elem.(*schema.Schema)
 			} else {
 				values, _ := schemaListToFlatValues(d.Get(key).([]interface{}), value.Elem.(*schema.Resource))

